### PR TITLE
BAAS-22176: add mem usage estimation for map objects

### PR DESCRIPTION
**Pull request by Gabri3l**
As a drive-by I defined a sample rate that is easier to manage across all estimates, I think eventually we could pass it as a param to make our tests simpler.

### CODE REVIEW

File: array.go [hunk 1]
```diff
@@ -580,7 +580,7 @@ func (a *arrayObject) estimateMemUsage(ctx *MemUsageContext) (estimate uint64, e
 	if arrayLen == 0 {
 		return memUsage, nil
 	}
-	sampleSize := arrayLen / 10
+	sampleSize := int(math.Floor(float64(arrayLen) * SampleRate))
```
Comment (nickpoindexter) [inline array.go]: [opt] Make this a function so it's only defined in one place.

File: builtin_map.go [hunk 2]
```diff
@@ -94,6 +95,40 @@ func (mo *mapObject) exportToMap(dst reflect.Value, typ reflect.Type, ctx *objec
...
+	var samplesVisited, memUsage uint64
+	totalItems := mo.m.size
+	if totalItems == 0 {
+		return memUsage, nil
+	}
+	sampleSize := int(math.Floor(float64(totalItems) * SampleRate))
+
+	i := 0
+	for item := mo.m.iterFirst; item != nil && i < totalItems; item = item.iterNext {
+		if i%sampleSize != 0 {
```
Comment (nickpoindexter) [inline builtin_map.go]: [q] Do we need to use `i` here? I think we can remove `i` and just use `samplesVisited`.

File: mem_context.go [hunk 3]
```diff
@@ -95,6 +95,8 @@ func (m *MemUsageContext) MemUsageLimitExceeded(memUsage uint64) bool {
 
 var (
 	ErrMaxDepth = errors.New("reached max depth")
+	// SampleRate represents the percentage of samples we pick up to estimate mem usage
+	SampleRate = 0.1
```
Comment (nickpoindexter) [inline mem_context.go]: [q] I know we don't want to spend too much time estimating large items, but wondering if this is high enough and maybe 20% is better given our mem issues.

Comment (Gabri3l) [reply mem_context.go]: yeah I think 20% would be good!

**Review by kpatel71716 — APPROVED** [review]
LGTM! Nothing from outside of Nick's comments. Interested to see the impact, if any, of 20% sampling but I'm down to go to 20

**Comment by Gabri3l** [comment]
Sorry I introduced a bunch of changes but the commits should help review this, it's mostly a refactoring for the changes in `mem_context.go`, I realized I'd rather be able to set the sample rate from baas instead of having to make changes here every time.

File: mem_context.go [hunk 4]
```diff
@@ -84,6 +87,9 @@ func NewMemUsageContext(
 			// number of obj props beyond which we should estimate mem usage
 			return objPropsLen > objPropsLenThreshold
 		},
+		ComputeSampleStep: func(totalItems int) int {
```
Comment (nickpoindexter) [inline mem_context.go:90]: [opt] I know this is the pattern we have, but we could just write this as a receiver method.

File: memory_test.go [hunk 5]
```diff
@@ -498,104 +498,104 @@ func TestMemArraysWithLenThreshold(t *testing.T) {
...
-			memLimit:  memUsageLimit,
-			expectedSizeDiff: SizeEmptyStruct + // Array overhead
-				(50 * SizeNumber) + (50 * SizeBool) + // (450) size of property values
-				// stack difference in going from
-				//	[..other, []]
-				//  to
-				//	[..other, true, 50]
-				+1,
-		},
+		// {
```
Comment (nickpoindexter) [inline memory_test.go]: [q] Should we add this back in or remove it?

File: memory_test.go [hunk 6]
```diff
@@ -498,104 +498,104 @@ func TestMemArraysWithLenThreshold(t *testing.T) {
...
-			expectedSizeDiff: SizeEmptyStruct + // Array overhead,
-				(100 * SizeNumber) + // size of property values
+				(10 * SizeNumber) + (10 * SizeBool) + // (450) size of property values
 				// stack difference in going from
 				//	[..other, []]
 				//  to
 				//	[..other, true, 50]
 				+1,
 		},
+		// {
```
Comment (nickpoindexter) [inline memory_test.go]: [q] same as above, but this looks like a new test that maybe is no longer necessary.

**Review by nickpoindexter — APPROVED** [review]
LGTM! Just some test cleanup

File: mem_context.go (hunk 4 above)
Comment (nickpoindexter) [reply mem_context.go:90]: Now that I'm thinking about this never mind, it's outside of scope for the PR

File: memory_test.go (hunk 5 above)
Comment (Gabri3l) [reply memory_test.go]: ahah I commented out by accident =/

**Pull request merged**